diff --unified --recursive --text spud.orig/diamond/setup.py.in spud.new/diamond/setup.py.in
--- spud.orig/diamond/setup.py.in	2019-06-27 12:40:30.612009000 +0100
+++ spud.new/diamond/setup.py.in	2019-06-27 12:46:37.601846206 +0100
@@ -46,30 +46,30 @@
 plugin_dirs = [dir for dir in os.listdir('plugins') if os.path.isdir(os.path.join('plugins', dir)) and dir[0] != '.']
 plugin_data_files = []
 if sys.platform == 'darwin' and packaging:
-	for plugin in plugin_dirs:
-  		plugin_data_files.append(("./plugins/" + plugin,
-		  glob.glob('plugins/' + plugin + '/*.py')))
+  for plugin in plugin_dirs:
+      plugin_data_files.append(("./plugins/" + plugin,
+      glob.glob('plugins/' + plugin + '/*.py')))
 else:
-	for plugin in plugin_dirs:
-	  if prefix is None:
-		plugin_data_files.append(("@prefix@/share/diamond/plugins/" + plugin,
-		  glob.glob('plugins/' + plugin + '/*.py')))
-	  else:
-		plugin_data_files.append((prefix + "/share/diamond/plugins/" + plugin,
-		  glob.glob('plugins/' + plugin + '/*.py')))
+  for plugin in plugin_dirs:
+    if prefix is None:
+      plugin_data_files.append(("@prefix@/share/diamond/plugins/" + plugin,
+        glob.glob('plugins/' + plugin + '/*.py')))
+    else:
+      plugin_data_files.append((prefix + "/share/diamond/plugins/" + plugin,
+        glob.glob('plugins/' + plugin + '/*.py')))
 
 # Now parse the GUI directories
 gui_data_files = []
-if sys.platform == 'darwin' and packaging	:
+if sys.platform == 'darwin' and packaging  :
   gui_data_files.append(("./gui",
     ["gui/gui.glade", "gui/diamond.svg", "gui/diamond.png"]))
 else:
-	if prefix is None:
-	  gui_data_files.append(("@prefix@/share/diamond/gui",
-		["gui/gui.glade", "gui/diamond.svg"]))
-	else:
-	  gui_data_files.append((prefix + "/share/diamond/gui",
-		["gui/gui.glade", "gui/diamond.svg"]))
+  if prefix is None:
+    gui_data_files.append(("@prefix@/share/diamond/gui",
+    ["gui/gui.glade", "gui/diamond.svg"]))
+  else:
+    gui_data_files.append((prefix + "/share/diamond/gui",
+    ["gui/gui.glade", "gui/diamond.svg"]))
 
 # We now have all the information we need; run setup.
 setup(
